src/ServiceControl.Monitoring/Http/Diagrams/TimingsAggregator.cs

```diff
@@ -34,11 +43,35 @@ public static IEnumerable<MonitoredEndpoint> AggregateIntoLogicalEndpoints(Proce
             }
         }
 
-        static MonitoredEndpointTimings AggregateTimings(List<TimingsStore.EndpointInstanceTimings> timings)
+
+        internal IEnumerable<MonitoredEndpointInstance> AggregateDataForLogicalEndpoint(string endpointName)
+        {
+            var now = DateTime.UtcNow;
+
+            var processingTime = processingTimeStore.GetTimings(now).Where(pt => pt.Id.EndpointName == endpointName).ToList();
+            var criticalTime = criticalTimeStore.GetTimings(now).Where(ct => ct.Id.EndpointName == endpointName).ToList();
+
+            var instanceIds = criticalTime.Concat(processingTime).Select(i => i.Id).Distinct();
+
+            foreach (var instanceId in instanceIds)
+            {
+                var monitoredInstance = new MonitoredEndpointInstance
+                {
+                    Id = instanceId.InstanceId,
+                    Name = instanceId.InstanceName,
+                    ProcessingTime = AggregateTimings(processingTime.Where(pt => pt.Id == instanceId).ToList()),
+                    CriticalTime = AggregateTimings(criticalTime.Where(ct => ct.Id == instanceId).ToList())
```

Comment:
These will perform reference equality, even though we've overridden `Object.Equals`.

We should override the equality operator or use `.Equals()` here https://msdn.microsoft.com/ru-ru/library/ms173147(v=vs.80).aspx